# Input and Output Formats
* For all input and configuration files in Biofilter, lines beginning with the symbol "`#`" will be ignored. This is useful for placing comments within input files that will not be a part of the analysis.

## Configuration Files (not required)
This example configuration file was generated by the `--report-configuration` option, with
everything else left at default values:
```
# Biofilter configuration file
# generated Sat, 20 Apr 2024 11:03:08
# Biofilter version 3.0.0 (2024-04-20)
# LOKI version 3.0.0 (2024-05-21)
REPORT_CONFIGURATION                            yes
REPORT_REPLICATION_FINGERPRINT                  no
REPORT_GENOME_BUILD                             yes
REPORT_GENE_NAME_STATS                          no
REPORT_GROUP_NAME_STATS                         no
ALLOW_UNVALIDATED_SNP_POSITIONS                 yes
ALLOW_AMBIGUOUS_SNPS                            no
ALLOW_AMBIGUOUS_KNOWLEDGE                       no
REDUCE_AMBIGUOUS_KNOWLEDGE                      no
REPORT_LD_PROFILES                              no
USER_DEFINED_FILTER                             no
GENE_IDENTIFIER_TYPE                            -
ALLOW_AMBIGUOUS_GENES                           no
GROUP_IDENTIFIER_TYPE                           -
ALLOW_AMBIGUOUS_GROUPS                          no
COORDINATE_BASE                                 1
REGIONS_HALF_OPEN                               no
REGION_POSITION_MARGIN                          0
REGION_MATCH_PERCENT                            100.0
REGION_MATCH_BASES                              0
MAXIMUM_MODEL_COUNT                             0
ALTERNATE_MODEL_FILTERING                       no
ALL_PAIRWISE_MODELS                             no
MAXIMUM_MODEL_GROUP_SIZE                        30
MINIMUM_MODEL_SCORE                             2
SORT_MODELS                                     yes
PARIS_P_VALUE                                   0.05
PARIS_ZERO_P_VALUES                             ignore
PARIS_ENFORCE_INPUT_CHROMOSOME                  yes
PARIS_PERMUTATION_COUNT                         1000
PARIS_BIN_SIZE                                  10000
PARIS_DETAILS                                   no
QUIET                                           no
VERBOSE                                         no
PREFIX                                          biofilter
OVERWRITE                                       no
STDOUT                                          no
REPORT_INVALID_INPUT                            no
PARIS                                           no
ALLOW_DUPLICATE_OUTPUT                          no
```

Any option which can be used on the command line can also be used in a configuration file. Each option must appear as the first item on a line, and any arguments to that option must be separated by `whitespace` (any number of tabs or spaces).

If an argument to an option must itself contain spaces (for example a multi-word gene or group identifier), the argument may be enclosed with `“`double quotes`”` to prevent the additional words in the argument from being interpreted as a separate arguments. If an argument must itself contain double quotes, they must be escaped with a backslash, `\”`like so`\”`.

There is also one extra option which may only be used in a configuration file: `INCLUDE`. This option requires one or more filename arguments and causes Biofilter to read each specified file as an additional
configuration file. Included files are processed in full before any other options in the original configuration file. For example, if file A includes file B and both files specify the same option, then the option’s setting or value from file A will always override the one from file B, even if it appears before the `INCLUDE` instruction. Included configuration files may also include further files; there is no limit to this recursion, except that any loops (i.e. A includes B which includes A) will raise an error.

