<img src="images/ritchie-lab-logo-2022-v3-horizontal-transparent.png" width="250">

# Welcome to the Ritchie Lab's *readthedocs* Site!
To learn more about our lab, please visit our [lab website](https://ritchielab.org/).

We are in the process of update our lab's software documentation to follow the readthedocs convention and plan to host software manuals and tutorials here in one convenient place.

# Brief Overview of Ritchie Lab Software 
The Ritchie lab has been developing bioinformatics tools and computational research methods for over 20 years. Below is a brief overivew of our tools:

## Biofilter and LOKI
Biofilter provides a convenient single interface for accessing multiple publicly available human genetic data sources that have been compiled in the supporting database of the Library of Knowledge Integration (LOKI). Information within LOKI includes genomic locations of SNPs and genes, as well as known relationships among genes and proteins such as interaction pairs, pathways and ontological categories.

July 2024: *Documentation for Biofilter and LOKI is actively being migrated to and updated on readthedocs*

## ATHENA
The Analysis Tool for Heritable and Environmental Network Associations (ATHENA) is a software package that combines statistical and biological variable selection methods with machine learning modeling techniques to identify complex prediction models that can include non-linear interactions and different types of high-throughput data.

More info here: [https://ritchielab.org/software/athena-downloads-1](https://ritchielab.org/software/athena-downloads-1)

## BioBin
BioBin is a standalone command line application and collapsing method that uses prior biological knowledge from a prebuilt database. While it was specifically developed to investigate rare variant burden in traditional genetic trait studies, BioBin can be used to apply multiple levels of burden testing and is useful for exploring the natural distribution of rare variants in ancestral populations.

More info here: [https://ritchielab.org/software/biobin-download-1](https://ritchielab.org/software/biobin-download-1 )

## PhenoGram
With PhenoGram researchers can create chomosomal ideograms annotated with lines in color at specific base-pair locations, or colored base-pair to base-pair regions, with or without other annotation. PhenoGram allows for annotation of chromosomal locations and/or regions with shapes in different colors, gene identifiers, or other text. PhenoGram also allows for creation of plots showing expanded chromosomal locations, providing a way to show results for specific chromosomal regions in greater detail.

More info here: [https://ritchielab.org/software/phewas-view-download](https://ritchielab.org/software/phewas-view-download) 

## PheWAS-View
We have developed the software PheWAS-View for visually integrating PheWAS results, including information about the SNPs, relevant genes, phenotypes, and the interrelationships between phenotypes, that exist in PheWAS. As a result both the fine grain detail as well as the larger trends that exist within PheWAS results can be elucidated.

More info here: [https://ritchielab.org/software/phewas-view-download](https://ritchielab.org/software/phewas-view-download)